### PR TITLE
TF-1861 Implement worker manager

### PR DESCRIPTION
**Pull request by dab246**
### Issue

#1861 

### Resolved

- [x] Config work manager on Android/iOS
- [x] Add abstracts class  for work manager

### CODE REVIEW

**Review by hoangdat — COMMENTED** [review]
Cool. Can you write the test? And don't forget to end up with an ADR

File: ios/Runner/AppDelegate.swift [hunk 1]
```diff
@@ -3,23 +3,32 @@ import Flutter
...
-        FlutterLocalNotificationsPlugin.setPluginRegistrantCallback { (registry) in
-            GeneratedPluginRegistrant.register(with: registry)
-        }
-
         if #available(iOS 10.0, *) {
             UNUserNotificationCenter.current().delegate = self as UNUserNotificationCenterDelegate
         }
         
-        GeneratedPluginRegistrant.register(with: self)
+        UIApplication.shared.setMinimumBackgroundFetchInterval(TimeInterval(60*15))
```
[the remark this answers is not in the document]
Comment (dab246) [reply ios/Runner/AppDelegate.swift:27]: This ensures that the task is ran at most every 15 minutes. See more at [here](https://github.com/fluttercommunity/flutter_workmanager/blob/main/IOS_SETUP.md)

File: lib/main.dart [hunk 2]
```diff
@@ -17,11 +18,16 @@ void main() async {
...
-    await MainBindings().dependencies();
-    await HiveCacheConfig().setUp();
-    await HiveCacheConfig.initializeEncryptionKey();
-    await Executor().warmUp();
-    await AppUtils.loadEnvFile();
+
+    await Future.wait([
+       MainBindings().dependencies(),
+       HiveCacheConfig().setUp(),
+       HiveCacheConfig.initializeEncryptionKey(),
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/main.dart:25]: Oh, That was my mistake, it must be done after `HiveCacheConfig().setUp()` is completed.

File: ios/Runner/Info.plist [hunk 3]
```diff
@@ -61,8 +61,13 @@
 	<key>UIBackgroundModes</key>
 	<array>
 		<string>fetch</string>
-		<string>remote-notification</string>
+        <string>remote-notification</string>
```
[the remark this answers is not in the document]
Comment (dab246) [reply ios/Runner/Info.plist:64]: It aligned properly in Xcode and Android Studio. IMO, it's just different rendering on github.

File: lib/features/offline_mode/scheduler/worker.dart [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import 'package:core/utils/platform_info.dart';
+import 'package:equatable/equatable.dart';
+import 'package:tmail_ui_user/features/offline_mode/scheduler/worker_type.dart';
+
+/// Equivalent to the task or work that needs to be done in the background
+class Worker with EquatableMixin {
+
+  final String id;
+  final WorkerType type;
+  final String data;
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/offline_mode/scheduler/worker.dart:11]: Our input can only be supported as primitive values and their lists. 
Screenshot 2023-05-23 at 16 43 21

So to pass an object I have to convert it to `json` string and assign it to `data`.

Screenshot 2023-05-23 at 16 40 42

We can distinguish different tasks such as (Insert, delete, sent email) through the `type` variable passed in.
Screenshot 2023-05-23 at 16 41 56

**Comment by dab246** [comment]
> Cool. Can you write the test? And don't forget to end up with an ADR

The tasks are done in `background` mode so writing unit-tests is quite difficult. 
We will write the ADR for its configuration and deployment.

File: lib/features/offline_mode/scheduler/worker.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (hoangdat) [reply lib/features/offline_mode/scheduler/worker.dart]: Scare about the json seriable 😨

File: lib/features/offline_mode/scheduler/worker.dart [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import 'package:core/utils/platform_info.dart';
+import 'package:equatable/equatable.dart';
+import 'package:tmail_ui_user/features/offline_mode/scheduler/worker_type.dart';
+
+/// Equivalent to the task or work that needs to be done in the background
+class Worker with EquatableMixin {
+
+  final String id;
+  final WorkerType type;
+  final Map<String, dynamic> data;
```
Comment (hoangdat) [inline lib/features/offline_mode/scheduler/worker.dart:11]: 👍

**Review by nqhhdev — APPROVED** [review]

**Pull request merged**